### PR TITLE
fix: ensure forwarding constructors have constraints

### PR DESCRIPTION
**Pull request by richardapeters**
ensure forwarding constructors have constraints that prevent copying / moving objects of the same type

### CODE REVIEW

File: infra/stream/InputStream.hpp [hunk 1]
```diff
@@ -63,13 +63,53 @@ namespace infra
         StreamErrorPolicy& errorPolicy;
     };
 
+    template<class Parent, class TheReader>
```
Comment (rjaegers) [inline infra/stream/InputStream.hpp:66]: Can we find a better name for TheReader? E.g. ReaderType?

File: infra/stream/OutputStream.hpp [hunk 2]
```diff
@@ -82,13 +82,53 @@ namespace infra
         StreamErrorPolicy& errorPolicy;
     };
 
+    template<class Parent, class TheWriter>
```
Comment (rjaegers) [inline infra/stream/OutputStream.hpp:85]: Same comment for TheWriter, WriterType?

**Review by rjaegers — APPROVED** [review]
Can you check the remaining six critical code-smells? It seems they are easy to fix.

File: infra/stream/InputStream.hpp [hunk 3]
```diff
@@ -63,13 +63,53 @@ namespace infra
...
+    class InputStreamWithReader
+        : private detail::StorageHolder<TheReader, InputStreamWithReader<Parent, TheReader>>
+        , public Parent
+    {
+    public:
+        InputStreamWithReader();
+        template<class Arg>
+        explicit InputStreamWithReader(Arg&& arg, std::enable_if_t<!std::is_same_v<InputStreamWithReader, std::remove_cv_t<std::remove_reference_t<Arg>>>, std::nullptr_t> = nullptr);
```
Comment (rjaegers) [inline infra/stream/InputStream.hpp:74]: I created a wrapper in util/Compatibility.hpp for enable_if_t. Do we still want to provide C++11 compatibility or do we drop that completely?

**Review by rjaegers — COMMENTED** [review]
Should this be a breaking change? Since it changes the interface of  the stream readers and writers?

**Comment by richardapeters** [comment]
> Should this be a breaking change? Since it changes the interface of the stream readers and writers?

Technically it breaks the interface, but not in a way that should matter. We prevent the incorrect constructor being called

File: infra/stream/InputStream.hpp (hunk 1 above)
Comment (richardapeters) [reply infra/stream/InputStream.hpp:66]: Sounds good, I'll apply it

File: infra/stream/OutputStream.hpp (hunk 2 above)
Comment (richardapeters) [reply infra/stream/OutputStream.hpp:85]: Same answer, sounds good, I'll apply it :)

**Pull request merged**